src/main/java/ch/njol/skript/expressions/ExprBanData.java

```diff
@@ -0,0 +1,226 @@
+import ch.njol.skript.doc.*;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.ExpressionType;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.skript.lang.util.SimpleExpression;
+import ch.njol.skript.util.Timespan;
+import ch.njol.util.Kleenean;
+import com.destroystokyo.paper.profile.PlayerProfile;
+import io.papermc.paper.ban.BanListType;
+import org.bukkit.BanEntry;
+import org.bukkit.Bukkit;
+import org.bukkit.OfflinePlayer;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+
+@Name("Ban Details")
+@Description("Returns data about a player or IP ban. " +
+	"The source can be any string, but it's usually the player's name. " +
+	"The expiration date won't return a value if the ban is permanent.")
+@Examples({
+	"set {_player} to offlineplayer(\"Notch\", false)",
+	"set {_expiration} to the date {_player}'s ban expires",
+	"set {_time.left} to difference between now and {_expiration}",
+	"set {_reason} to the reason {_player} was banned",
+	"send \"There is %{_time.left}% before %{_player}% gets unbanned! They were banned for '%{_reason}%'\" to player"
+})
+@Since("INSERT VERSION")
+@RequiredPlugins("Spigot 1.20.1+")
+public class ExprBanData extends SimpleExpression<Object> {
+
+	static {
+		if (Skript.methodExists(BanEntry.class, "remove"))
+			Skript.registerExpression(ExprBanData.class, Object.class, ExpressionType.SIMPLE,
+				"[the] date %offlineplayer/string% was banned",
+				"[the] date of %offlineplayer/string%'s ban",
+				"[the] ban date of %offlineplayer/string%",
+
+				"[the] source of %offlineplayer/string%'s ban",
+				"[the] ban source of %offlineplayer/string%",
+
+				"[the] date %offlineplayer/string%'s ban expires",
+				"[the] expiration date of %offlineplayer/string%'s ban",
+				"[the] ban expiration date of %offlineplayer/string%",
+
+				"[the] reason %offlineplayer/string% was banned",
+				"[the] reason for %offlineplayer/string%'s ban",
+				"[the] ban reason of %offlineplayer/string%"
+		);
+	}
+
+	private BanEntryType entryType;
+	private Expression<Object> banTargetExpr; // offline player or string (ip)
+
+	@Override
+	public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		entryType = BanEntryType.fromInt(matchedPattern);
+		banTargetExpr = (Expression<Object>) expressions[0];
+		return true;
+	}
+
+	@Override
+	protected Object @Nullable [] get(Event event) {
+		if (banTargetExpr == null)
+			return null;
+
+		Object banTarget = banTargetExpr.getSingle(event);
+		BanEntry<?> banEntry;
+
+		if (banTarget instanceof String ipTarget)
+			banEntry = getBanEntry(ipTarget);
+		else if (banTarget instanceof OfflinePlayer playerTarget)
+			banEntry = getBanEntry(playerTarget);
+		else
+			return null;
+
+		if (banEntry == null)
+			return null;
+
+		return switch (entryType) {
```

Comment:
The returned array type needs to match `getReturnType()` (that is, returning Object arrays is not safe)